presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndex.java

```diff
@@ -0,0 +1,223 @@
+    private static final Envelope ZERO_ENVELOPE = new Envelope(0.0, 0.0, 0.0, 0.0);
+
+    private final LongArrayList addresses;
+    private final List<Type> types;
+    private final List<Integer> outputChannels;
+    private final List<List<Block>> channels;
+    private final STRtree rtree;
+    private final Optional<JoinFilterFunction> filterFunction;
+    private final long memorySizeInBytes;
+
+    private static final class GeometryWithAddress
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(GeometryWithAddress.class).instanceSize();
+        private final OGCGeometry ogcGeometry;
+        private final long address;
+
+        private GeometryWithAddress(OGCGeometry ogcGeometry, long address)
+        {
+            this.ogcGeometry = ogcGeometry;
+            this.address = address;
+        }
+
+        public OGCGeometry getOgcGeometry()
+        {
+            return ogcGeometry;
+        }
+
+        public long getAddress()
+        {
+            return address;
+        }
+
+        public long getEstimatedMemorySizeInBytes()
+        {
+            return INSTANCE_SIZE + GeometryUtils.getEstimatedMemorySizeInBytes(ogcGeometry);
+        }
+    }
+
+    public PagesSpatialIndex(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            int geometryChannel,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.types = types;
+        this.outputChannels = outputChannels;
+        this.channels = requireNonNull(channels, "channels is null");
+        this.rtree = buildRTree(geometryChannel);
+        this.filterFunction = filterFunctionFactory.map(factory -> factory.create(session.toConnectorSession(), addresses, channels));
+        this.memorySizeInBytes = computeMemorySizeInBytes();
+    }
+
+    @Nullable
+    private STRtree buildRTree(int geometryChannel)
+    {
+        STRtree rtree = new STRtree();
+
+        for (int position = 0; position < addresses.size(); position++) {
+            long pageAddress = addresses.getLong(position);
+            int blockIndex = decodeSliceIndex(pageAddress);
+            int blockPosition = decodePosition(pageAddress);
+
+            Block block = channels.get(geometryChannel).get(blockIndex);
+            if (block.isNull(blockPosition)) {
+                continue;
+            }
+
+            Slice slice = block.getSlice(blockPosition, 0, block.getSliceLength(blockPosition));
+            OGCGeometry ogcGeometry = deserialize(slice);
+            if (ogcGeometry.isEmpty()) {
+                continue;
+            }
+
+            rtree.insert(getEnvelope(ogcGeometry), new GeometryWithAddress(ogcGeometry, pageAddress));
+        }
+
+        if (rtree.isEmpty()) {
+            return null;
```

Comment:
maybe you could have a `PagesSpatialIndex::Factory` which in case of empty `rtree` would return dummy `PagesSpatialIndex` implementation?

Reply:
I'll try that.